# Lab 5: Logistic Regression - Explanatory Model

## Google Drive Setup

In [28]:
from google.colab import drive
import os

# Mount your google drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [29]:
# **Important**: Set your current working directory (CWD) to the folder
# containing your notebook and testing files

notebook_path = '/content/drive/MyDrive/Lab5_student'
os.chdir(notebook_path)

# Make sure your notebook and "tests" folder exist in this path
print(os.getcwd())

/content/drive/MyDrive/Lab5_student


## Initialize Otter Grader

In [30]:
%pip install -q otter-grader

In [31]:
# Initialize Otter
import otter
grader = otter.Notebook()

## Business and Data Understanding

The Credit Approval dataset consists of actual credit card applications and their status for either approval or denial. The predictors have been **anonymized** to protect confidential information about the applicants.

This dataset is interesting because there is a good mix of attributes -- continuous, nominal with small numbers of values, and nominal with larger numbers of values. There are also a few missing values.

##Additional Variable Information
- V1:	b, a.
- V2:	continuous.
- V3:	continuous.
- V4:	u, y, l, t.
- V5:	g, p, gg.
- V6:	c, d, cc, i, j, k, m, r, q, w, x, e, aa, ff.
- V7:	v, h, bb, j, n, z, dd, ff, o.
- V8:	continuous.
- V9:	t, f.
- V10:	t, f.
- V11:	continuous.
- V12:	t, f.
- V13:	g, p, s.
- V14:	continuous.
- V15:	continuous.
- **Target: approved/denied**

## Install and import required packages

In [32]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

import hashlib
def get_hash(num):
    """Helper function for assessing correctness"""
    return hashlib.md5(str(num).encode()).hexdigest()

# Interactive shell. Enable if needed
#from IPython.core.interactiveshell import InteractiveShell
#InteractiveShell.ast_node_interactivity = "all"

## Load the data and explore data

In [33]:
data_df = pd.read_csv('https://raw.githubusercontent.com/rul110/datasets/refs/heads/main/credit_approval.csv')

**Perform these steps and then answer Q1**


(1) Show the first 5 rows of the dataset

In [34]:
data_df.head()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,Target
0,b,30.83,0.000,u,g,w,v,1.25,t,t,1,f,g,202.0,0,Approved
1,a,58.67,4.460,u,g,q,h,3.04,t,t,6,f,g,43.0,560,Approved
2,a,24.50,0.500,u,g,q,h,1.50,t,f,0,f,g,280.0,824,Approved
3,b,27.83,1.540,u,g,w,v,3.75,t,t,5,t,g,100.0,3,Approved
4,b,20.17,5.625,u,g,w,v,1.71,t,f,0,f,s,120.0,0,Approved


(2) Display dataset size, the variable names, count of non-null values, and datatypes

In [35]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      678 non-null    object 
 1   V2      678 non-null    float64
 2   V3      690 non-null    float64
 3   V4      684 non-null    object 
 4   V5      684 non-null    object 
 5   V6      681 non-null    object 
 6   V7      681 non-null    object 
 7   V8      690 non-null    float64
 8   V9      690 non-null    object 
 9   V10     690 non-null    object 
 10  V11     690 non-null    int64  
 11  V12     690 non-null    object 
 12  V13     690 non-null    object 
 13  V14     677 non-null    float64
 14  V15     690 non-null    int64  
 15  Target  690 non-null    object 
dtypes: float64(4), int64(2), object(10)
memory usage: 86.4+ KB


(3) Find the number of observations in each category of the target variable


In [36]:
data_df.Target.value_counts()

,count
Target,
Denied,383
Approved,307


### Q1

**Which statement is INCORRECT?**

(A) This dataset has one target variable and 16 predictors, which includes 6 numerical variables and 9 categorical variables.

(B) The numerical variables have different scales.

(C) Six variables have small percentages of missing values

(D) The target variable is slightly imbalanced distributed, with about 60% denied observations.

_Points:_ 2.5

In [37]:
...


# (1) Show the first 5 rows of the dataset


# (2) Display dataset size, the variable names, count of non-null values, and datatypes


# (3) Find the number of observations in each category of the target variable


Q1_Answer = 'C'


print(f"\nYour Answer: {Q1_Answer}")


Your Answer: C


In [38]:
grader.check("Q1")

Q1 results: All test cases passed!

### Handle Missing Values

Remember we talked the following methods to handle missing values:
- Drop the variable
- Drop observations with missing values
- Impute missing values
- Treat missing as its own informative category

**For variables V1 (categorical; missing 22) and V2 (numerical; missing 22), what methods can be applied?**

### Q2

For simplicity, let's drop observations with missing values

**Drop observations with missing values**

_Points:_ 2.5

In [39]:
data_df = data_df.dropna(axis= 0)

# Drop observations with missing values

# Show dataset size after dropping observations with missing values
data_size = data_df.shape


print(f"dataset size after dropping observations with missing values: {data_size}")

dataset size after dropping observations with missing values: (653, 16)


In [40]:
grader.check("Q2")

Q2 results: All test cases passed!

## Explanatory Logistic Regression Model

### Prepare Data

### Q3

Write code to complete the following tasks:
- Separate variables into predictors (X) and target (y)
- Dummy encode categorical variables in X (with drop_first = True, dtype = int)
- Map target values: "denied" to label 0 and "approved" to 1 (i.e., positive class)
- Calculate the percentage of 1s (i.e., approved) to check label balance

In [41]:
data_df.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'Target'],
      dtype='object')

In [65]:
predictors = ['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15']
X = data_df[predictors]
y = data_df['Target']

# Dummy encoding
X = pd.get_dummies(X, drop_first = True, dtype = int)

# map labels to 0/1
label_dict = {'Denied': 0, 'Approved': 1}
y = y.map(label_dict)
percentage_of_approved = y.mean() * 100


print(f"Number of predictors in X: {X.shape}")
print(f"Value distribution in y: {y.value_counts()}")
print(f"% of Approved observations: {percentage_of_approved :.2f}%")

Number of predictors in X: (653, 37)
Value distribution in y: Target
0    357
1    296
Name: count, dtype: int64
% of Approved observations: 45.33%


In [66]:
grader.check("Q3")

Q3 results: All test cases passed!

### Model Development

### Q4

Write code to do the following:

- Fit a logistic regression model using (X,y) with arguments `max_iter = 5000, random_state = 123`
- Print intercept and coefficients

In [46]:

logit_model = LogisticRegression(max_iter= 5000, random_state= 123)
logit_model.fit(X,y)

intercept = logit_model.intercept_
coef_df = pd.DataFrame({"Predictor": X.columns,
                        "Coefficient": logit_model.coef_[0]})


print(f"Intercept: {intercept}")
print(f"Coefficients: :\n{coef_df}")


Intercept: [-2.45702561]
Coefficients: :
   Predictor  Coefficient
0         V2     0.002882
1         V3    -0.024818
2         V8     0.096044
3        V11     0.122107
4        V14    -0.001985
5        V15     0.000536
6       V1_b     0.003495
7       V4_u    -0.198530
8       V4_y    -0.460236
9      V5_gg     0.647963
10      V5_p    -0.460236
11      V6_c     0.072444
12     V6_cc     0.950853
13      V6_d     0.189878
14      V6_e     0.342429
15     V6_ff    -0.963735
16      V6_i    -0.586424
17      V6_j    -0.325121
18      V6_k    -0.466059
19      V6_m    -0.137131
20      V6_q     0.168919
21      V6_r    -0.004119
22      V6_w     0.501482
23      V6_x     1.508964
24     V7_dd     0.053218
25     V7_ff    -0.268698
26      V7_h     0.349579
27      V7_j     0.469200
28      V7_n     0.673169
29      V7_o    -0.044351
30      V7_v     0.019712
31      V7_z    -0.650771
32      V9_t     3.247004
33     V10_t     0.435374
34     V12_t    -0.108363
35     V13_p    -0.0652

In [47]:
grader.check("Q4")

Q4 results: All test cases passed!

 ### Q5

**Which of the following statements regarding the fitted model is correct?**

A. When all predictors equal zero, the probability that Y=1 is approximately $\frac{1}{1+e^{2.46}}$.

B. A one-unit increase in V3 increases the probability that Y=1 by 2.49%.

C. A one-unit increase of V2 is associated with 0.0028 increase on the probability that Y=1.

D. A one-unit increase in V11 multiplies the probability that Y=1 by $e^
{0.122}$
 .

_Points:_ 2.5

In [48]:
...

Q5_Answer = 'A'


print(f"Q5 Answer: {Q5_Answer}")


Q5 Answer: A


In [49]:
grader.check("Q5")

Q5 results: All test cases passed!

### Model Evaluation

### Q6

Write code to evaluate model performance on training data as follows:
- Accuracy
- Precision/Recall/F1
- Display the actual label, predicted label, and predicted probability of the first 5 observations

In [62]:
# Predict label probability
y_hat_prob = logit_model.predict_proba(X)[:, 1]

# Predict label
y_hat = logit_model.predict(X)

# Create a dataframe with three columns: actual label, predicted label, predicted probability
pred_df = pd.DataFrame({"Actual": y,
                        "Predicted_Label": y_hat,
                        "Predicted_Prob": y_hat_prob})

accuracy = (pred_df. Actual == pred_df.Predicted_Label).mean()

# Print classification report
print(classification_report(y, y_hat))


print(f"Training accuracy: {accuracy :.4f}\n")

print(f"Prediction of first 5 observations: :\n{pred_df.head()}")


              precision    recall  f1-score   support

           0       0.92      0.86      0.89       357
           1       0.85      0.91      0.88       296

    accuracy                           0.88       653
   macro avg       0.88      0.89      0.88       653
weighted avg       0.89      0.88      0.88       653

Training accuracy: 0.8836

Prediction of first 5 observations: :
   Actual  Predicted_Label  Predicted_Prob
0       1                1        0.814801
1       1                1        0.944974
2       1                1        0.768221
3       1                1        0.905418
4       1                1        0.763000


In [63]:
grader.check("Q6")

Q6 results: All test cases passed!

---

To double-check your work, the cell below will rerun all of the autograder tests.

In [67]:
grader.check_all()

Q1 results: All test cases passed!

Q2 results: All test cases passed!

Q3 results: All test cases passed!

Q4 results: All test cases passed!

Q5 results: All test cases passed!

Q6 results: All test cases passed!